### DISCUSSION TITLE
How do I type something that isn't already built in to the framework?

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [typed-ember/glint#348] [Q&A] [answered]
Per this suggestion: https://github.com/emberjs/rfcs/issues/822#issuecomment-1140775399

I was trying to figure out how to type class-based resources.

after my class definition, I tried the example from the linked comment above:
```ts
export interface Resource<T extends ArgsWrapper> extends InstanceType<
  HelperLike<{
    Args: { Named: T['named']; Positional: T['positional'] };
    // this is the instance of the Resource -- which I also don't know how to get
    // Return: IDontKnowHowToGetTheValueTypeOfAResourceButItGoesHere
    Return: Resource<T> // how to make this work with sub classes?
  }>
> {}
```
now,  I'm working with the class definition here because folks can do:

```ts
class MyResource extends Resource {
  @tracked foo;
}

<template>
  {{#let (MyResource whateverArgs) as |instance|
    {{instance.foo}}
  {{/let}}
</template>
```
In JS/TS, I have some type-help via a static method (this is for context, rather than directly related to the issue as resources are usable in both templates and js/ts):
```ts
class {
  instance = MyResource.from(() => whateverArgs);
}
```

but anyway, back to `HelperLike`, I'm getting this error:
```
 TS2344: Type 'HelperLike<{ Args: { Named: T["named"]; Positional: T["positional"]; }; Return: new (...args: any) => any; }>' does not satisfy the constraint 'new (...args: any) => any'.
  Cannot assign an abstract constructor type to a non-abstract constructor type.
```

So, it seems `HelperLike` cannot be passed to `InstanceType`.

I can remove the `InstanceType` and the red squigglies go away, _however_, I get a very hard to read error in one of the utilities consming this class:

```
 TS2344: Type 'TrackedArrayMap<Element, MapTo>' does not satisfy the constraint 'Resource<ArgsWrapper>'.
  Type 'Invokable<(args_0: NamedArgs<Element, MapTo>, args_1: Element[]) => Resource<{ positional: PositionalArgs<Element>; named: NamedArgs<Element, MapTo>; }>>' is not assignable to type 'Invokable<(...args: [named: EmptyObject | Record<string, any>, ...positional: unknown[]] | [named: EmptyObject | Record<string, any>]) => Resource<ArgsWrapper>>'.
    Type '(args_0: NamedArgs<Element, MapTo>, args_1: Element[]) => Resource<{ positional: PositionalArgs<Element>; named: NamedArgs<Element, MapTo>; }>' is not assignable to type '(...args: [named: EmptyObject | Record<string, any>, ...positional: unknown[]] | [named: EmptyObject | Record<string, any>]) => Resource<ArgsWrapper>'.
      Types of parameters 'args_0' and 'args' are incompatible.
        Type '[named: EmptyObject | Record<string, any>, ...positional: unknown[]] | [named: EmptyObject | Record<string, any>]' is not assignable to type '[NamedArgs<Element, MapTo>, Element[]]'.
          Type '[named: EmptyObject | Record<string, any>, ...positional: unknown[]]' is not assignable to type '[NamedArgs<Element, MapTo>, Element[]]'.
            Target requires 2 element(s) but source may have fewer.
src/util/map.ts (182:39)

182   let resource = TrackedArrayMap.from<TrackedArrayMap<Element, MapTo>>(
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

`from` is inherited from `Resource`, which looks like
```ts
  static from<Instance extends Resource<ArgsWrapper>>(
    this: (new (...args: unknown[]) => Instance),
    context: object,
    thunk?: Thunk | (() => unknown)
  ): Instance {
```

:thinking: this is a lot, I made a [TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgCQKYBsyqgGWAa1TgF84AzKCEOAcgAEBzdYAOxgHoZVx0BDLmgG4AUMNQAPSLDisuUMrwDGRAIJQGAZwDqUXmCxREwuHEgbgMYBBa90AfgBccAK4t8LCAHcWAbQC6IiY2IKgAJo5wAEqoitChADwaMFCsDAA0cLwsAJ4AfCLEYpLQ8DDZWHAAKgAWrvhwALzGcOzscAByEHBqms0APnAAFIMAlI25cP4j-UOj43AAbhDAodMmraZ8rJlQutktphDmlta2O70mA8NjDRPRAI7OwFBh8T3auvrYuT40ZhZWGzoGh+NYtNpgLYsODBMLnDQza7zB5PF4Jd46PQGH40WGhEFgjYAIwgMGqMN4IVCmRY1P+JyB8IAdCy4ETnPBPNV9p4IM50NTsny4IpqhAjkQycANHY4A5nGBzKEwl4WAplA5EXNbt11B8sdgRiJRBIpPBFHwNBooqgNHyoMp4pU4BIuLTrRjPgZGrrNJivlAJghmkl+MBFORKCB4gBJFihljKF3iN2ha3RO3OB2oN56-3Y3KDZomKUaJyDFioTxDFlM3h6px1DzeKbzOMJ5QjNLFkXWLgppwQIkAKxiMG7JhLtTcERqdTgV21EybqrBIyc7ZgWSTwcncBeMCz0IADJlrSvvGe4Jvt6hBOCZDxuKg2GHrHAqBYuKFmoVmrF42SZxFBgaBBlVbBGzcZsWDGXdJw2DRUBgAB5bxsEGUsMggqAjV-UQTBACBQmAMhsjsQZ6UBWwnEqX4qNOYE-AyPFaN+PECScJYVgKE1imkWRsHVIgM3tR1nVdF8019fUAx9T0DUDZNUwREw0EwbA8EIeJ4JMd4nCQdpKTCNjcWM-EAjgAAFI4AUY0yGKBEESECSdokPKAWCcUSs3E-JfyDP9hFaELQrC8KIsiqLopi1pgraABVSxmDKXt42cEBUjgMkiF4EkFlQIozRkNghKUIgAFkDVCNQ9niKqvlCSoICDZofFYZVxCcFgMqJbA-CcBqsCaiBXIWWxnFtcsbgmIawma-xXPQF8GDJbreuwVyfAAZWyEASXQJkLGwfhoD8UYNzkU6oHq6rmv8oKygqGzjmo9B3niABRH0LxYCYGkmT7-ACYRBPkcqOnMj7voB36Mgzfl4Fh6DVVawi9HLDBnzYJxPpmm0NER3jgoAKhJ4wSbgOgwBScauAp9gipKEVLWtSpdEUQgat2Xhsgar7lpCNgfpR7wMjmkaRfcVHmkk90CbEnNdMOV77Os2yGVsaHBZfGB-L3VjIapbXsfHOAJfu1ziFyZpQEhU3rQl2reduxr7uEeCNkqDB0DZ7bsuqfg4E8IhFCyHY9gAWg6iRI6UZQrTamOuphDaoAG827tGgi4AAYhAPQfUrastFQXh8H5z6deFgAyD8RzHcWs8LPDmiIkiyMGHxQn4XgM5euygRNoW9YyJAC7AEgnCM429QF02m7dlq4J7DZPxgb98Oacb0Em60AaXSZa1LEHmgYZC4GWlhVuqOZlYPI84GPK2c52vaDqOq7QPTi7ry-6BXbDXukYPcD9PJPyvL9K8MZ-43Qti1e8a9MobzCPhIKTNpBkFcCBQEcAJ7zxHlLGCi9mpENRkWEwyokiUGyHlZag4G4gQnB+MADIywgMnD3Lc015hV1NotHsE9MbVxgLjER+MGrNRfivEwy14BIC4bwDIE8SA+ggKwwEGhjTrDaAAMWAOILKpZ9yoEeM8W0VR-YABYmQAHYMhISIB4Lg1ptg+WzEyCAZBmhyJMZmbMPp2ZKC5s7PmegmQUCoE6DmISeZhLAAQ3WJCWqFh7FQ5IEBaFEmWswkwh9layIvi8UiwA4QA0UaMbRe4TEeWhEgRyNFJjFLIKU0IzEKRUgMngwupBiCuRMIUEwrcdEugKlAbIUpr4ul9kQde34hhWUoOIbIYIwHQheP45QvEgA)

So I guess assuming I'm using HelperLike correctly, how do I read this new error? is there anything I could do about my types to flatten it? or to make it less verbose?
I get lost mid-way though.

**Comment by dfreeman** [comment, marked as answer]
> So I guess assuming I'm using HelperLike correctly,

You're not 😉 
The snippet I gave you with `InstanceType` was using it correctly, and linked to an example of that exact usage working in practice in `@glint/ember-environment-ember-loose`.

What version of TS are you using? `InstanceType` has accepted abstract constructors since [this PR](https://github.com/microsoft/TypeScript/commit/e10a32591ce57b788a23ddbf3ead470b28c8a820) that landed in 4.3. Note that writing `extends InstanceType<HelperLike<...>>` in the playground you linked doesn't produce an error.

**Reply by NullVoxPopuli** [reply]
oh! I was using typescript 4.2. Gonna try dropping support for 4.2, 4.3, and 4.4, since all this work I'm doing is in prep for a major anyway :upside_down_face:

**Reply by NullVoxPopuli** [reply]
success!!! 
Silly 4.2 

thank you!

**Reply by NullVoxPopuli** [reply]
Well, my error about map remains. but at least InstanceType is working

**Reply by NullVoxPopuli** [reply]
More specific questions to come later
